force-app/main/default/classes/LeadTriggerHandlerTest.cls

```diff
@@ -0,0 +1,319 @@
+        List<Contact> contacts = [SELECT Id, Email, AccountId FROM Contact];
+        Set<String> contactEmails = new Set<String>();
+        Map<String, Id> emailToContactIdMap = new Map<String, Id>();
+        
+        for (Contact c : contacts) {
+            if (c.Email != null) {
+                contactEmails.add(c.Email);
+                emailToContactIdMap.put(c.Email, c.Id);
+            }    
+        }
+
+        List<Lead> leads = [SELECT Id, Email FROM Lead WHERE Email IN :contactEmails];
+
+        Test.startTest();
+        LeadTriggerHandler.handleLeadAutoConvert(leads);
+        Test.stopTest();
+
+        List<Lead> updatedLeads = [SELECT Id, Email, IsConverted, ConvertedContactId FROM Lead WHERE Id IN :leads];
+
+        for (Lead updatedLead : updatedLeads) {
+            Id contactId = emailToContactIdMap.get(updatedLead.Email);
+            
+            if (contactId != null) {
+                Assert.isTrue(updatedLead.IsConverted, 'Lead should have been converted for: ' + updatedLead.Email);
+                Assert.areEqual(contactId, updatedLead.ConvertedContactId, 'Lead was not converted to the correct Contact for: ' + updatedLead.Email);
+            } else {
+                Assert.isTrue(!updatedLead.IsConverted, 'Lead should not have been converted for: ' + updatedLead.Email);
+            }
+        }
+    }
+    
+    @isTest
+    static void HandleLeadAutoConvert_nullEmails() {
+        // Create Accounts
+        Account account = new Account(Name = 'Test Account');
+        insert account;
+
+        // Create Contacts with unique emails
+        List<Contact> contacts = new List<Contact>();
+        for (Integer i = 0; i < 5; i++) {
+            contacts.add(new Contact(
+                FirstName = 'ContactFirstName' + i,
+                LastName = 'ContactLastName' + i,
+                Email = 'email' + i + '@example.com',
+                AccountId = account.Id
+            ));
+        }
+        insert contacts;
+
+        // Create Leads, with some having null email addresses
+        List<Lead> leads = new List<Lead>();
+        for (Integer i = 0; i < 5; i++) {
+            leads.add(new Lead(
+                FirstName = 'LeadFirstName' + i,
+                LastName = 'LeadLastName' + i,
+                Company = 'Company' + i,
+                Email = Math.mod(i, 2) == 0 ? null : 'email' + i + '@example.com' // alternate between null and valid email
+            ));
+        }
+        insert leads;
+
+        // Run the trigger handler method
+        Test.startTest();
+        LeadTriggerHandler.handleLeadAutoConvert(leads);
+        Test.stopTest();
+
+        // Query the leads again to check conversion status
+        List<Lead> updatedLeads = [SELECT Id, IsConverted, Email FROM Lead WHERE Id IN :leads];
+        
+        // Verify that leads with null emails have not been converted
+        for (Lead ld : updatedLeads) {
+            if (ld.Email == null) {
```

Comment:
Start practicing using the string class to check for other bad values like empty space or ''